scripts/sudo.ps1

```diff
@@ -1,87 +1,94 @@
+<#
+	.SYNOPSIS
+		Runs a scriptblock, command or application as an elevated process using sudo.exe
+	.DESCRIPTION
+		Wraps sudo.exe to add functionality for running PowerShell scripts or commands in an elevated process.
+
+		When running a scriptblock or PowerShell command, a new copy of PowerShell is run with sudo and an EncodedCommand.
+		When running a native application, the command itself is run with sudo.
+
+		This script DOES NOT (currently) support piping input to the elevated command.
+#>
+
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT License.
+[CmdletBinding(DefaultParameterSetName = "Script")]
+param(
+	# A scriptblock to run in an elevated process
+	[Parameter(Mandatory, Position = 0, ParameterSetName = "Script")]
+	[scriptblock]$ScriptBlock,
 
-# open question - Should -NoProfile be used when invoking PowerShell
-BEGIN {
-	if ($__SUDO_TEST -ne $true) {
-		$SUDOEXE = "sudo.exe"
-	}
-	else {
-		if ($null -eq $SUDOEXE) {
-			throw "variable SUDOEXE has not been set for testing"
-		}
-	}
+	# Run PowerShell with the -NoProfile switch
+	[switch]$NoProfile,
 
-	if ([Environment]::OSVersion.Platform -ne "Win32NT") {
+	# A command or application to run in an elevated process
+	[Parameter(Mandatory, Position = 0, ParameterSetName = "Command")]
+	[string]$Command,
+
+	# Arguments to pass to the command or application
+	[Parameter(Position = 1, ParameterSetName = "Command", ValueFromRemainingArguments)]
+	[Alias("Args")]
+	[PSObject[]]$ArgumentList
+)
+begin {
+	if ($IsLinux -or $IsMacOS) {
 		throw "This script works only on Microsoft Windows"
 	}
 
-	if ($null -eq (Get-Command -Type Application -Name "$SUDOEXE" -ErrorAction Ignore)) {
-		throw "'$SUDOEXE' cannot be found."
+	if ($__SUDO_TEST -ne $true) {
+		$Env:SUDOEXE = "sudo.exe"
+	} elseif (!$Env:SUDOEXE) {
+			throw "Environment variable SUDOEXE has not been set for testing"
 	}
 
-	$psProcess = Get-Process -id $PID
-	if (($null -eq $psProcess) -or ($psProcess.Count -ne 1)) {
-		throw "Cannot retrieve process for '$PID'"
+	if (!(Get-Command -Type Application -Name $Env:SUDOEXE -ErrorAction Ignore)) {
+		throw "Env:SUDOEXE is set to '$Env:SUDOEXE' but it cannot be found."
 	}
 
-	$thisPowerShell = $psProcess.MainModule.FileName
-	if ($null -eq $thisPowerShell) {
-		throw "Cannot determine path to '$psProcess'"
+	$thisPowerShell = (Get-Process -Id $PID).MainModule.FileName
```

Comment:
I want to point out that I dramatically simplified this, just because there isn't really any way this can go wrong. $PID is a _constant_ in PowerShell and can't changed, so this is always going to return the full path to the executable.